src/ir/effects.h

```diff
@@ -143,10 +145,37 @@ class EffectAnalyzer {
     return globalsRead.size() || readsMemory || readsHeap || isAtomic || calls;
   }
 
-  bool hasSideEffects() const {
+  bool hasNonTrapSideEffects() const {
     return localsWritten.size() > 0 || danglingPop || writesGlobalState() ||
-           trap || throws || transfersControlFlow();
+           throws || transfersControlFlow();
   }
+
+  bool hasSideEffects() const { return trap || hasNonTrapSideEffects(); }
+
+  // Check if there are side effects, and they are of a kind that cannot be
+  // removed by optimization passes.
+  //
+  // The difference between this and hasSideEffects is subtle, and only related
+  // to trapsNeverHappen - if trapsNeverHappen then any trap we see is removable
+  // by optimizations. In general, you should call hasSideEffects, and only call
+  // this method if you are certain that it is a place that would not perform an
+  // unsafe transformation with a trap. Specifically, if a pass calls this
+  // and gets the result that there are no unremovable side effects, then it
+  // must either
+  //
+  //  1. Remove any side effects present, if any, so they no longer exists.
+  //  2. Keep the code exactly where it is.
+  //
+  // If instead of 1&2 a pass kept the side effect and also reordered the code
+  // with other things, then that could be bad, as the side effect might have
+  // been behind a condition that avoids it occurring.
+  //
+  // TODO: Go through the optimizer and use this in all places that do not move
+  //       code around.
+  bool hasUnremovableSideEffects() const {
```

Comment:
I assume the common case of usage will be `if (!hasUnremovableSideEffects) { /* remove side effects */ }`. Would it be worth eliminating that double negative by changing this to `canRemoveSideEffects` or even `canRemove`?

Reply:
Not necessarily, I think. One of the two usages already present is the opposite,
```
// Check for various things that make us fail.
if (some check) return;
if (hasUnremovableSideEffects()) return;
if (other check) return;
// All good, do the opt here.
```
Basically we replace `hasSideEffects()` with `hasUnremovableSideEffects()`. I think keeping those two methods similar is easier to keep in one's head, personally.